### PR TITLE
Add torch._dynamo.config.fail_on_cache_limit_hit

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136767

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/_dynamo/exc.py [hunk 1]
```diff
@@ -188,6 +188,13 @@ class IncorrectUsage(Exception):
     pass
 
 
+# TODO: I'm a little uncertain about what error classification we should have
+# for this.  This is potentially a user error, but regressions in
+# specialization in PyTorch proper could also trigger this problem
+class FailOnCacheLimitHit(Exception):
```
Comment (albanD) [inline torch/_dynamo/exc.py:194]: Maybe subclass of RuntimeError? Not sure indeed.

**Review by albanD — APPROVED** [review]
SGTM

File: torch/_dynamo/exc.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/exc.py:194]: Yeah I really don't know... maybe we should bother the error classification guys some more haha

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**